### DISCUSSION TITLE
Ocaml target

### DISCUSSION THREAD

**Discussion by bernardro** [antlr/antlr4#3760] [Ideas]
Hi, I would like to start work on an Ocaml target. I have reviewed the page on new Antlr target creation. Besides that does anyone have any ideas, advice, links, etc?

**Comment by ericvergnaud** [comment]
Hey,

great to hear that.

We have a requirement that all targets are compatible ie. the same grammar behaves the same.
To help with this, we have a strong focus on CI, with the exact same tests running across all targets.
So a good sequence could be as follows:
 - create a test target (similar to the existing ones for other targets)
 - implement CI scripts for CircleCI Linux and/or GitHub Mac and/or GitHub Windows
 - start with simple lexer tests, then complex lexer tests
 - continue with simple parser tests

Give up any form of social life, family life or mental health.
Feel free to ask questions, don’t feel offended if you don’t get a response quickly.

That said, welcome to the community !

Eric

**Reply by KvanTTT** [reply]
I'd move CI scripts to the end, they aren't required if other items are not completed.

Creating a new test target is very easy with the new test infrastructure, especially after https://github.com/antlr/antlr4/pull/3752

**Comment by KvanTTT** [comment]
Hi, I'm not sure @parrt will integrate new targets, see https://github.com/antlr/antlr4/pull/3736#issuecomment-1152981109 Also I'm not sure OCaml is in high demand.

**Comment by parrt** [comment]
Hi, yeah, I don't think it makes sense to add this target to the main repository although you are of course free to work on one for your own use.

**Comment by bernardro** [comment]
Thanks for the ideas and warnings. My approach would be to avoid reinventing the wheel if possible - so I may need to suffer multiple passes through multiple tools to get a stable working end product. Still sifting through the open source offerings for these kinds of tasks though - and there is a ton.

Another one that caught my eye is the sitter one and Facebook's evolution of it. It does trigger mixed feelings because its AI.. and we don't have explicit up-front control over how it makes decisions. Fallacies aside, it might be the way to go to iterate and test a more formally verified tool. Antlr is battle tested however, especially for the early grammars and targets.

As for maintenance - I wonder if there is a way to break up and restructure the Antlr repo so that owners of targets and grammars are responsible for keeping their repos, grammars, etc. up to date. I am new to Antlr so these are just my rumblings at this point.

**Reply by parrt** [reply]
We originally had to decide between multiple repositories and a single one. We might've even tried separate ones. We currently do this for php and it complicates our lives for testing etc..

**Reply by KvanTTT** [reply]
Maybe it makes sense to use submodules, but a single repository is more maintainable indeed.
